docs/src/errortypes.md

```diff
@@ -0,0 +1,264 @@
+└──────────────
+```
+
+### How to fix
+This error is unique in that idiomatic Julia code may still lead to this error. For example, in the `pos_after_tab` function above, if the input vector does not have a `'\t'` byte, `p` will be `nothing`, and a `MethodError` will be thrown when `nothing + 1` is attempted.
+However, in many situations, the possibility of such a `MethodError` is not a mistake, but rather an idiomatic way of erroring.
+
+There are different possiblities to address this kind of error. Let's take the `pos_after_tab` example:
+
+If you actually _could_ expect `p` to legitimately be `nothing` for valid input (i.e. the input could lack a `'\t'` byte), then your function should be written to take this edge case into account:
+```julia
+julia> function pos_after_tab(v::AbstractArray{UInt8})
+    p = findfirst(isequal(UInt8('\t')), v)
+    if p === nothing # handle the nothing case
+        return nothing
+    else
+        return p + 1
+    end
+end;
+
+julia> @report_call pos_after_tab(codeunits("a\tb"))
+No errors detected
+```
+
+By adding the `if p === nothing` check, the compiler will know that the type of `p` must be `Nothing` inside the `if` block, and `Int` in the `else` block. This way, the compiler knows a `MethodError` is not possible, and the error will disappear.
+
+If you expect a `'\t'` byte to always be present, such that `findfirst` always should return an `Int` for valid input, you can add a typeassert in the function to assert that the return value of `findfirst` must be, say, an `Integer`. Then, the compiler will know that if the typeassert passes, the value returned by `findfirst` cannot be `nothing` (and hence in this case must be `Int`):
+
+```julia
+julia> function pos_after_tab(v::AbstractArray{UInt8})
+    p = findfirst(isequal(UInt8('\t')), v)::Integer
+    p + 1
+end
+
+julia> @report_call pos_after_tab(codeunits("a\tb"))
+No errors detected
+```
+
+The code will still error at runtime due to the typeassert if `findfirst` returns `nothing`, but JET will no longer detect it as an error, because the programmer, by adding the typeassert, explicitly acknowledge that the compiler's inference may not be precise enough, and helps the compiler.
+
+Note that adding a typeassert also improves code quality:
+* The programmer's intent to never observe `nothing` is communicated clearly
+* After the typeassert passes, `p` is inferred to be `Int` instead of a union, and this more precise type inference generates more efficient code.
+* More precise inference reduces the risk of invalidations from the code, improving latency.
+
+A special case occurs when loading `Union`-typed fields from structs.
+Julia does not realize that loading the same field multiple times necessarily returns the same object. Hence, in the following example:
+
+```julia
+julia> struct Foo
```

Comment:
For immutable structs, this "error" will no longer be reported after `VERSION ≥ v"1.10.0-DEV.25"` (xref: https://github.com/aviatesk/JET.jl/pull/421). Might be best to use `mutable struct` here.